pkg/server/grpc/authenticator.go

```diff
@@ -58,5 +78,68 @@ func (cca *clientCertAuthenticator) authenticate(ctx context.Context) *user {
 	}
 
 	ids := pki.ExtractIDs(chains[0][0].Extensions)
-	return &user{ids}
+	return &user{
+		authSource: authSourceClientCertificate,
+		identities: ids,
+	}
+}
+
+// An authenticator that extracts identity from JWT. The JWT is requied to be
+// transmitted using the "Bearer" authencitation scheme.
+type idTokenAuthenticator struct {
+	verifier *oidc.IDTokenVerifier
+}
+
+func newIDTokenAuthenticator(aud string) (*idTokenAuthenticator, error) {
+	provider, err := oidc.NewProvider(context.Background(), idTokenIssuer)
+	if err != nil {
+		return nil, err
+	}
+
+	verifier := provider.Verifier(&oidc.Config{ClientID: aud})
+	return &idTokenAuthenticator{verifier}, nil
+}
+
+func (ja *idTokenAuthenticator) authenticate(ctx context.Context) *user {
+	bearerToken := extractBearerToken(ctx)
+	if bearerToken == "" {
+		glog.Warning("no bearer token exists")
+
+		return nil
+	}
+
+	idToken, err := ja.verifier.Verify(context.Background(), bearerToken)
+	if err != nil {
+		glog.Warningf("failed to verify the ID token (error %v)", err)
+
+		return nil
+	}
+
+	return &user{
+		authSource: authSourceIDToken,
+		identities: []string{idToken.Subject},
+	}
+}
+
+func extractBearerToken(ctx context.Context) string {
+	md, ok := metadata.FromContext(ctx)
+	if !ok {
+		glog.Warning("no metadata is attached")
+		return ""
+	}
+
+	authHeader, exists := md[httpAuthHeader]
+	if !exists {
+		glog.Warning("no HTTP authorization header exists")
+		return ""
+	}
+
+	for _, value := range authHeader {
+		if strings.HasPrefix(value, bearerTokenPrefix) {
+			return strings.TrimPrefix(value, bearerTokenPrefix)
```

Comment:
Is it possible to have two bearer tokens?

Reply:
I feel this is rarely used. I'm looking into RFCs to see whether it's even allowed.

Reply:
Please ping back this thread once you are done.

Reply:
Dig more about this - though I can't find any RFC that forbids multiple authorization headers, the current grpc-go doesn't support multiple. See here: https://github.com/grpc/grpc-go/blob/master/transport/http2_client.go#L346

Reply:
sg